### DISCUSSION TITLE
USB Serial port to CalTopo w/Heltec V3 Issue

### DISCUSSION THREAD

**Discussion by DECLLC** [meshtastic/Discussions#120] [Firmware]
Hello,

I'm trying to get real-time data from a Heltec V3 in to CalTopo. I have downloaded and installed the local version of CalTopo.

I have edited the topo.config file:

sarsoft.location.aprs.local.enabled=true
sarsoft.location.serial.tty.usbserial-0001=9600,8,1,0

log4i.logger.aprsengine.name=org.sarsoft.location.service.APRSLocalEngine
log4i.logger.aprsengine.level=INFO

This is on Mac OS X Sequoia 15.1.1.

I don't see anything in the CalTopo console/debug window, except when I unplug the Heltec: 

> 2024-12-10 18:26:25 APRSLocalEngine [ERROR] Unhandled exception
> com.fazecast.jSerialComm.SerialPortInvalidPortException: Unable to create a serial port object from the invalid port descriptor: /dev/cu.usbserial-0001
> 	at com.fazecast.jSerialComm.SerialPort.getCommPort(SerialPort.java:549) ~[desktop.jar:0.9.10]
> 	at org.sarsoft.location.service.APRSLocalEngine.portDiscovery(APRSLocalEngine.java:78) ~[desktop.jar:0.9.10]
> 	at org.sarsoft.location.service.APRSLocalEngine.lambda$start$0(APRSLocalEngine.java:54) ~[desktop.jar:0.9.10]
> 	at java.lang.Thread.run(Thread.java:840) ~[?:?]
> Caused by: java.io.IOException
> 	at com.fazecast.jSerialComm.SerialPort.getCommPort(SerialPort.java:534) ~[desktop.jar:0.9.10]
> 	... 3 more

I have tried viewing the serial port in Chrome, and I see regular logs/debugging, but I don't seem to see anything specific to locations from other devices, notably a LilyGo T-Echo:

> INFO | 02:22:22 337 [Router] Received position from=0xa1ebbde6, id=0x4bc4d2b9, portnum=3, payloadlen=45
> 
> DEBUG | 02:22:22 337 [Router] POSITION node=a1ebbde6 l=45 lat=33XX22023 lon=-118XX76906 msl=-7 hae=0 geo=0 pdop=394 hdop=0 vdop=0 siv=7 fxq=0 fxt=0 pts=0 time=1733883739
> 
> DEBUG | 02:22:22 337 [Router] Current RTC quality: Net. Ignore time of RTC quality of Net
> 
> INFO | 02:22:22 337 [Router] updatePosition REMOTE node=0xa1ebbde6 time=1733883739 lat=33XX22023 lon=-118XX76906
> 
> DEBUG | 02:22:22 337 [Router] Node status update: 2 online, 36 total
> 
> DEBUG | 02:22:22 337 [Router] Send location with precision 32
> 
> INFO | 02:22:22 337 [Router] Strip low RTCQuality (2) time from position
> 
> INFO | 02:22:22 337 [Router] Position reply: time=0 lat=33XX22025 lon=-118XX75717

**Comment by DECLLC** [comment]
Small update, I now have data coming in to the CalTopo console - this seems to work from the LilyGo T-Echo hardware, but not from the Heltec V3. I'm not sure if it is a baud rate issue, or something else, as Chome seems to work to connect to the console, but screen and CalTopo does not.

However, regardless of the serial port settings in the Mestastic app I can't seem to get Mestastic local to parse the data.

I have tried NEMA, and serial.mode.caltopo. I am seeing the output, but not sure if there's a way to get the software to plot it as a "locator" in their terms. 

Example:

> INFO  | 16:25:42 703 [GPS] updatePosition LOCAL pos@675a2d85 time=1733963142 lat=33XX24278 lon=-118XX77935 alt=-12

**Comment by SamoJS** [comment]
Hi 

You're using the Console debug output. Try using a UART, by configuring the RX TX pins from the Meshtastic Radio Configuration app. For Heltec v3 I use Pins RX 47 and TX 48 and BAUD of 9600. To read off these pins I use an MKR 1010 wifi, or ESP32.  Any board with Serial and Serial1 is super easy....

Below is the code I use to read from the Heltec v3...
The code runs on the MKR 1010.
Super easy to run a jumper from the pins you configured on the Heltec to the serial1 pins of the MKR 1010 or an ESP32 for instance.
In this example the output of the Heltec appears on the console (usb) of the MKR 1010. 

Heltec   -         MKR 1010
TX pin (48) -> RX pin
RX pin (47) -> TX pin

```
/*
  Multiple Serial test

  Receives from the main serial port, sends to the others.
  Receives from serial port 1, sends to the main serial (Serial 0).

  This example works only with boards with more than one serial like Arduino Mega, Due, Zero etc.

  The circuit:
  - any serial device attached to Serial port 1
  - Serial Monitor open on Serial port 0

  created 30 Dec 2008
  modified 20 May 2012
  by Tom Igoe & Jed Roach
  modified 27 Nov 2015
  by Arturo Guadalupi

  This example code is in the public domain.

  https://www.arduino.cc/en/Tutorial/BuiltInExamples/MultiSerialMega
*/


void setup() {
  // initialize both serial ports:
  Serial.begin(9600);
  Serial1.begin(9600);
}

void loop() {
  // read from port 1, send to port 0:
  if (Serial1.available()) {
    int inByte = Serial1.read();
    Serial.write(inByte);
  }

  // read from port 0, send to port 1:
  if (Serial.available()) {
    int inByte = Serial.read();
    Serial1.write(inByte);
  }
}

```
I hope that makes sense.
